File: spark-scala/spark-dataset/src/main/scala/com/mahfooz/spark/dataset/Datasets.scala
/*

Datasets are the foundational type of the Structured APIs.
Datasets are a strictly Java Virtual Machine (JVM) language feature that work only with Scala and Java.

Dataset<T> in Java and Dataset[T] in Scala.

  val flightsDF =spark.read.format("csv")
  .option("header", "true")
  .option("inferSchema", "true")
  .load("2015-summary.csv")

This is very import as encoder.
  import spark.implicits._

In fact, if you use Scala or Java, all "DataFrames" are actually Datasets of type Row.

Starting with the Spark 2.0 release, there is only one high-level abstraction called a Dataset, which has two flavors:
  a)  a strongly typed API
  b)  an untyped API

DataFrame is essentially a type alias for Dataset[Row], where a Row is a generic untyped JVM object.
Dataset APIs are available in only strongly typed languages such as Scala and Java.
Dataset is known as a data structure in Spark SQL that provides type safety and object-oriented interface.

Dataset:

Acts as an extension to DataFrame API.
Combines the features of DataFrame and RDD.
Serves as a functional programming interface for working with structured data.
Overcomes the limitations of RDD and DataFrames - the absence of automatic optimization in RDD and absence of compile-time type safety in Dataframes.

Features of Dataset
  Offers optimized query using **Tungsten**and Catalyst Query optimizer.
  These will be discussed in upcoming topics.
  Capable of analyzing during compile-time.
  Capable of converting type-safe dataset to untyped DataFrame using methods - toDS():Dataset[A],
  toDF():DataFrame and toDF(columnnames:String):DataFrame.
  Delivers high performance due to faster computation.
  Low memory consumption.
  Provides unified API for Java and Scala.

DataFrames is an organized form of distributed data into named columns that is similar to a table in a relational database.
Dataset is an upgraded release of DataFrame that includes the functionality of object-oriented programming interface, type-safe and fast.

In DataFrames, data is represented as a distributed collection of row objects.
In DataSets, data is represented as rows internally and JVM Objects externally.

After RDD transformation into dataframe, it cannot be regenerated to its previous form.
After RDD transformation into a dataset, it is capable of converting back to original RDD.

In dataframe, a runtime error will occur while accessing a column that is not present in the table. It does not provide compile-time type safety.
In dataset, compile time error will occur in the same scenario as dataset provides the compile-time type safety.

 */

package com.mahfooz.spark.dataset

import com.mahfooz.spark.dataset.model.Flight
import org.apache.spark.sql.SparkSession

object Datasets {

  def main(args: Array[String]): Unit = {

    val spark = SparkSession
      .builder()
      .master("local[*]")
      .appName(getClass.getSimpleName)
      .getOrCreate()

    val flightsDF = spark.read
      .parquet(args(0))

    import spark.implicits._

    val flights = flightsDF.as[Flight]

    flights
      .filter(flight_row => flight_row.ORIGIN_COUNTRY_NAME != "Canada")
      .map(flight_row => flight_row)
      .limit(5)
      .show()

    flights
      .take(5)
      .filter(flight_row => flight_row.ORIGIN_COUNTRY_NAME != "Canada")
      .map(
        fr => Flight(fr.DEST_COUNTRY_NAME, fr.ORIGIN_COUNTRY_NAME, fr.count + 5)
      )

    flights.show(2)
    println(flights.first.DEST_COUNTRY_NAME)
  }
}

File: spark-java/spark-architecture/src/main/java/spark/api/dataset/DataSets.java
/*

A Dataset is a new experimental interface added in Spark 1.6 that tries to provide the benefits of RDDs (strong typing, ability to use powerful
lambda functions) with the benefits of Spark SQL’s optimized execution engine.

A Dataset can be constructed from JVM objects and then manipulated using functional transformations (map, flatMap, filter, etc.).

The unified Dataset API can be used both in Scala and Java.

Python does not yet have support for the Dataset API, but due to its dynamic nature many of the benefits are already available
(i.e. you can access the field of a row by name naturally row.columnName).

Full python support will be added in a future release.

before Spark 2.0, the main programming interface of Spark was the Resilient Distributed Dataset (RDD).

After Spark 2.0, RDDs are replaced by Dataset, which is strongly-typed like an RDD, but with richer optimizations under the hood.

Spark’s primary abstraction is a distributed collection of items called a Dataset.

Dataset is known as a data structure in Spark SQL that provides type safety and object-oriented interface.

Dataset:

	Acts as an extension to DataFrame API.
	Combines the features of DataFrame and RDD.
	Serves as a functional programming interface for working with structured data.
	Overcomes the limitations of RDD and DataFrames - the absence of automatic optimization in RDD and absence of compile-time type safety in Dataframes.

Creating Datasets
	Dataset uses a specialized Encoder to serialize the objects for processing or transmitting over the network.
	
	val dsCaseClass = Seq(Person("John", 35)).toDS()
    dsCaseClass.show()
	
Dataset from RDD
A dataset can be created from RDD as shown:

    val ds=rdd.DS()
    ds.show()

Dataset from DataFrame
	
	val peopleDS = spark.read.json("examples/src/main/resources/data.json").as[Person]
	peopleDS.show()
 ===============================
 
 
Import all required packages
	
	import org.apache.spark.sql.SparkSession

Create a SparkSession object
	
	val spark= SparkSession.builder.appName("My Spark Application").master("local[*]").config("spark.sql.warehouse.dir", "/root/spark-warehouse").getOrCreate

Call the function spark.range to get the output starting from 5 to 50, with increments of 5
	
	val numDS = spark.range(5, 50, 5)

Display the value in DataSet
	numDS.show()

Call the function 'orderby' to reverse the order in DataSet and display the first 5 values
	
	numDS.orderBy(desc("id")).show(5)

Compute descriptive stats and display them
	numDS.describe().show()
	
 */
package spark.api.dataset;

public class DataSets {
}
